### PR TITLE
[19435] Apply thread settings

### PR DESCRIPTION
**Pull request by MiguelCompany**
## Description

This PR adds utility methods to create a std::thread with a custom name and thread settings.

Implementations of the method for all supported platforms are provided.

It also refactors the places where a thread is created to use the new proxy method, using the thread settings from the corresponding QoS.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [ ] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [ ] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 
    
- **N/A** Applicable backports have been included in the description.

## Reviewer Checklist
- [ ] The PR has a milestone assigned.
- [ ] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

File: src/cpp/utils/threading.hpp [hunk 1]
```diff
@@ -55,6 +64,66 @@ void set_name_to_current_thread(
...
+ *
+ * @param[in]  func      Functor with the logic to be run on the created thread.
+ * @param[in]  settings  Thread settings to apply to the created thread.
+ * @param[in]  name      Name (format) for the created thread.
+ * @param[in]  args      Additional arguments to complete the thread name.
+ *                       See @ref set_name_to_current_thread for details.
+ */
+template<typename Functor, typename ... Args>
+std::thread create_thread(
+        const Functor& func,
```
Comment (EduPonz) [inline src/cpp/utils/threading.hpp:90]:
```suggestion
        Functor func,
```

File: src/cpp/utils/threading.hpp [hunk 2]
```diff
@@ -55,6 +64,66 @@ void set_name_to_current_thread(
...
+        const fastdds::rtps::ThreadSettings& settings,
+        const char* name,
+        Args... args)
+{
+    return std::thread([&]()
```
Comment (EduPonz) [inline src/cpp/utils/threading.hpp:95]:
```suggestion
    return std::thread([=]()
```

**Review by EduPonz — CHANGES_REQUESTED** [review]
Very much a partial review

File: src/cpp/utils/threading/threading_win32.ipp [hunk 3]
```diff
@@ -70,6 +82,7 @@ static void configure_current_thread_affinity(
 void apply_thread_settings_to_current_thread(
         const fastdds::rtps::ThreadSettings& settings)
 {
+    configure_current_thread_priority(settings.priority);
```
Comment (EduPonz) [inline src/cpp/utils/threading/threading_win32.ipp:85]: If we cannot change sched policy in windows, he should state so in ThreadSettings.hpp

**Review by EduPonz — CHANGES_REQUESTED** [review]
Some remarks. Also we're still missing stack size. I guess we can do that in a follow up

File: src/cpp/utils/threading/threading_pthread.ipp [hunk 4]
```diff
@@ -13,8 +13,14 @@
 // limitations under the License.
 
 #include <pthread.h>
-#include <string.h>
 #include <stdio.h>
+#include <string.h>
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/utils/threading/threading_pthread.ipp]: 0c5d7d7

File: src/cpp/utils/threading/threading_pthread.ipp [hunk 5]
```diff
@@ -49,9 +55,61 @@ void set_name_to_current_thread(
...
+    else if((sched_class == SCHED_FIFO) ||
+            (sched_class == SCHED_RR))
+    {
+        //
+        // RT Policies use a different priority numberspace.
+        //
+        
+        param.sched_priority = sched_priority;
+        result = pthread_setschedparam(self_tid, sched_class, &param);
+    }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/utils/threading/threading_pthread.ipp:112]: a9f64d6, f1a299f, 393617a, 2383e86

File: include/fastdds/rtps/attributes/ThreadSettings.hpp [hunk 6]
```diff
@@ -59,16 +59,16 @@ struct RTPS_DllAPI ThreadSettings
     int32_t priority = 0;
 
     /**
-     * @brief The thread's core affinity.
+     * @brief The thread's affinity.
      *
-     * cpu_mask is a bit mask for setting the threads affinity to each core individually.
+     * On some systems, this is a bit mask for setting the threads affinity to each core individually.
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply include/fastdds/rtps/attributes/ThreadSettings.hpp:65]: 2f7faec

File: src/cpp/utils/threading/threading_win32.ipp (hunk 3 above)
Comment (MiguelCompany) [reply src/cpp/utils/threading/threading_win32.ipp:86]: 2f7faec

File: test/unittest/logging/LogTests.cpp [hunk 7]
```diff
@@ -664,6 +664,31 @@ TEST_F(LogTests, flush_n)
...
+    thr_settings.cpu_mask = 3;
+    thr_settings.scheduling_policy = SCHED_OTHER;
+    thr_settings.priority = 1;
+#endif // if defined(_POSIX_SOURCE)
+    Log::SetThreadConfig(thr_settings);
+
+    for (unsigned int i = 0; i < n_logs; i++)
+    {
+        EPROSIMA_LOG_INFO(TEST_THREADS, "Info message " << i);
+        std::this_thread::sleep_for(std::chrono::milliseconds(250));
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply test/unittest/logging/LogTests.cpp]: 68aac31

**Review by EduPonz — APPROVED** [review]

**Pull request merged**